test/gtest/ucs/test_arbiter.cc

```diff
@@ -816,3 +819,218 @@ UCS_TEST_F(test_arbiter_resched_from_dispatch, remove_and_resched) {
 
     ucs_arbiter_group_purge(&m_arb, &m_group2, purge_cb, NULL);
 }
+
+class test_arbiter_random_resched : public test_arbiter_resched_from_dispatch {
+public:
+    test_arbiter_random_resched();
+
+protected:
+    virtual ucs_arbiter_cb_result_t dispatch(ucs_arbiter_group_t *_group,
+                                             ucs_arbiter_elem_t *elem);
+
+    void do_test_loop(unsigned num_groups, unsigned elems_per_group,
+                      unsigned dispatch_per_group);
+
+private:
+    typedef struct {
+        ucs_arbiter_group_t  super;
+        unsigned             num_elems;
+    } arb_group_t;
+
+    void reset_counters();
+
+    void add_new_elem(arb_group_t *group, bool schedule);
+
+    void do_test(unsigned iteration_num, unsigned num_groups,
+                 unsigned elems_per_group, unsigned dispatch_per_group);
+
+    std::vector<arb_group_t> m_groups;
+    unsigned                 m_num_dispatch;
+    unsigned                 m_num_only;
+    unsigned                 m_num_added;
+    unsigned                 m_num_removed;
+    unsigned                 m_num_push_another;
+    unsigned                 m_num_next_group;
+    unsigned                 m_num_desched;
+    unsigned                 m_num_resched;
+};
+
+test_arbiter_random_resched::test_arbiter_random_resched()
+{
+    reset_counters();
+}
+
+void test_arbiter_random_resched::reset_counters()
+{
+    m_num_dispatch     = 0;
+    m_num_only         = 0;
+    m_num_added        = 0;
+    m_num_removed      = 0;
+    m_num_push_another = 0;
+    m_num_next_group   = 0;
+    m_num_desched      = 0;
+    m_num_resched      = 0;
+}
+
+void test_arbiter_random_resched::add_new_elem(arb_group_t *group, bool schedule)
+{
+    ucs_arbiter_elem_t *elem = new ucs_arbiter_elem_t;
+
+    ucs_arbiter_elem_init(elem);
+    ucs_arbiter_group_push_elem(&group->super, elem);
+    if (schedule) {
+        ucs_arbiter_group_schedule(&m_arb, &group->super);
+    }
+    ++group->num_elems;
+    ++m_num_added;
+}
+
+ucs_arbiter_cb_result_t
+test_arbiter_random_resched::dispatch(ucs_arbiter_group_t *_group,
+                                      ucs_arbiter_elem_t *elem)
+{
+    arb_group_t *group = ucs_derived_of(_group, arb_group_t);
+    arb_group_t *new_group;
+
+    ++m_num_dispatch;
+
+    /* Test ucs_arbiter_elem_is_only() */
+    if (group->num_elems == 1) {
+        ++m_num_only;
+        EXPECT_TRUE(ucs_arbiter_elem_is_only(&group->super, elem));
+    }
+
+    if (ucs::rand() % 2) {
+        /* Remove the current element.
+         * Must remove elements with higher probability than adding to avoid
+         * infinite loop.
+         */
+        if ((m_groups.size() > 1) && (ucs::rand() % 4) == 0) {
+            /* push the removed element to a random group which is not the
+             * current group */
+            new_group = &m_groups[ucs::rand() % (m_groups.size() - 1)];
+            if (new_group >= group) {
+               ++new_group;
```

Comment:
new_group will never point to last element since % is on (size-1) so it can be up to (size-2)
if i do like this:
```
if (new_group == group) {
    ++new_group;
}
```
then the group after current one will have non-uniformly higher probability to be selected